### ISSUE TITLE
Very High CPU Load but Low GPU Load when Training YOLOv3 - caused by transformer function and PrefetchTargetGenerator

### ISSUE THREAD

**Issue by ElectronicElephant** [dmlc/gluon-cv#1188]
First, really appreciate your work.

I have encountered with the problem that the CPU load is very high when training YOLOv3 on different machines.

## Environment
2 x E5-2630 v4 @ 2.20GHz / 2 x E5-2678 v3 @ 2.50GHz
Titan Xp / Titan RTX
Ubuntu 16.04
Python 3.6 / 3.7
mxnet-cu92mkl / mxnet-cu92 / mxnet-cu101mkl, etc

## Intel-MKL ?
At first, I thought it was a Intel-MKL-related issue. Thus, I set the environmental variables
```
export MKL_NUM_THREADS="1"
export MKL_DOMAIN_NUM_THREADS="MKL_BLAS=1"
export OMP_NUM_THREADS="1"
export MKL_DYNAMIC="FALSE"
export OMP_DYNAMIC="FALSE"
```
This saves some CPU load, but it's still very high.

## Issue
I have set `num_workers` to be exactly the number of my CPU cores. Thus, the CPU load is extremely high. (Load Average in htop > 60)

However, in `nvidia-smi`, I noticed that the load on GPUs is hardly above 70%. It will reach around 80% every 8 to 10 seconds, and stays at around 20% in other time.  However, ideally, the load on GPUs should always be above 90%.

The training speed was around 45 minutes per epoch on COCO dataset. It's good, but with such high CPU load, other jobs can hardly go on.

These all suggest me that the bottle-neck is the CPU workers.

I have tried to disable all the tricks in the transformer. This almost **double** the speed!

---

To sum up, I guess the problem lays in the two:

- YOLOV3PrefetchTargetGenerator
- Image Transformer

---

In the PrefetchTargetGenerator, I noticed
```
            # TODO(zhreshold): the number of valid gt is not a big number, therefore for loop
            # should not be a problem right now. Switch to better solution is needed.
```
and in the transformer function, there is a comment
```
        # generate training target so cpu workers can help reduce the workload on gpu
```

It's understandable that we should make GPUs and CPUs work in parallel to save time. However, the problem is that the load on CPUs is just too heavy. Do you have any advice?

**Comment by EletronicElephant** [comment]
For example, can I move the data-argumentation process to one of the GPUs?

Or, will replacing the functions with `mxnet.images` help?

**Comment by zhreshold** [comment]
The current workaround is to use DALI dataloader which is already provided in yolo training script.

Regarding you question:
Some transformations including detection target generation is not yet supported on GPU, the high CPU load is a known issue which we are working on but is long term.  
Note that disabling some transformations or data augmentation will help speed up the training but will cause model to perform worse.

**Comment by ElectronicElephant** [comment]
> 
> 
> The current workaround is to use DALI dataloader which is already provided in yolo training script.
> 
> Regarding you question:
> Some transformations including detection target generation is not yet supported on GPU, the high CPU load is a known issue which we are working on but is long term.
> Note that disabling some transformations or data augmentation will help speed up the training but will cause model to perform worse.

Hello, thank you for your reply. I do know that `DALI`, `Horovod`, and `AMP` are integrated into `ssd`, which boosts the speed by nearly 500%.

However, I don't see anything about `DALI` in the training script. `https://github.com/dmlc/gluon-cv/blob/master/scripts/detection/yolo/train_yolo3.py`. There are `AMP` and `Horovod`, but no `DALI`.

**Comment by zhreshold** [comment]
@Kh4L Did I miss something, I remember https://github.com/dmlc/gluon-cv/pull/986 should add DALI support as well right?

**Comment by Kh4L** [comment]
@zhreshold DALI integration was postponed, as it was not ready fully ready for YOLO detection a the time.

**Comment by ElectronicElephant** [comment]
@zhreshold  Okay, since there is no black magic to boost performance for the time being, personally, I'd suggest adding environmental variable `export OMP_NUM_THREADS="1"` by default. MMDetection also adopts it AFAIK.
